### Exploración básica de datos
Este

Cargamos las bibliotecas a utilizar

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

#### Cargamos el conjunto de datos

In [2]:
# Colocamos la ruta al conjunto de datos en una variable para facilitar el acceso
melbourne_file_path = "https://raw.githubusercontent.com/vbatiz/intro-ML/main/notebooks/datasets/melb_data.csv"
# Leemos los datos y los almacenamos en un dataframe llamado melbourne_data
melbourne_data = pd.read_csv(melbourne_file_path)
# Desplegamos un resumen de los datos del dataframe
melbourne_data.describe()

Rooms         Price      Distance      Postcode      Bedroom2  \
count  13580.000000  1.358000e+04  13580.000000  13580.000000  13580.000000   
mean       2.937997  1.075684e+06     10.137776   3105.301915      2.914728   
std        0.955748  6.393107e+05      5.868725     90.676964      0.965921   
min        1.000000  8.500000e+04      0.000000   3000.000000      0.000000   
25%        2.000000  6.500000e+05      6.100000   3044.000000      2.000000   
50%        3.000000  9.030000e+05      9.200000   3084.000000      3.000000   
75%        3.000000  1.330000e+06     13.000000   3148.000000      3.000000   
max       10.000000  9.000000e+06     48.100000   3977.000000     20.000000   

           Bathroom           Car       Landsize  BuildingArea    YearBuilt  \
count  13580.000000  13518.000000   13580.000000   7130.000000  8205.000000   
mean       1.534242      1.610075     558.416127    151.967650  1964.684217   
std        0.691712      0.962634    3990.669241    541.014538    37.273762   
min        0.000000      0.000000       0.000000      0.000000  1196.000000   
25%        1.000000      1.000000     177.000000     93.000000  1940.000000   
50%        1.000000      2.000000     440.000000    126.000000  1970.000000   
75%        2.000000      2.000000     651.000000    174.000000  1999.000000   
max        8.000000     10.000000  433014.000000  44515.000000  2018.000000   

          Lattitude    Longtitude  Propertycount  
count  13580.000000  13580.000000   13580.000000  
mean     -37.809203    144.995216    7454.417378  
std        0.079260      0.103916    4378.581772  
min      -38.182550    144.431810     249.000000  
25%      -37.856822    144.929600    4380.000000  
50%      -37.802355    145.000100    6555.000000  
75%      -37.756400    145.058305   10331.000000  
max      -37.408530    145.526350   21650.000000

#### Usando el mismo esquema cargue el conjunto de datos `train_homedata.csv`

In [3]:
homedata_file_path = "https://raw.githubusercontent.com/vbatiz/intro-ML/main/notebooks/datasets/train_homedata.csv"
home_data = pd.read_csv(homedata_file_path)

In [4]:
home_data.describe()

Id   MSSubClass  LotFrontage        LotArea  OverallQual  \
count  1460.000000  1460.000000  1201.000000    1460.000000  1460.000000   
mean    730.500000    56.897260    70.049958   10516.828082     6.099315   
std     421.610009    42.300571    24.284752    9981.264932     1.382997   
min       1.000000    20.000000    21.000000    1300.000000     1.000000   
25%     365.750000    20.000000    59.000000    7553.500000     5.000000   
50%     730.500000    50.000000    69.000000    9478.500000     6.000000   
75%    1095.250000    70.000000    80.000000   11601.500000     7.000000   
max    1460.000000   190.000000   313.000000  215245.000000    10.000000   

       OverallCond    YearBuilt  YearRemodAdd   MasVnrArea   BsmtFinSF1  ...  \
count  1460.000000  1460.000000   1460.000000  1452.000000  1460.000000  ...   
mean      5.575342  1971.267808   1984.865753   103.685262   443.639726  ...   
std       1.112799    30.202904     20.645407   181.066207   456.098091  ...   
min       1.000000  1872.000000   1950.000000     0.000000     0.000000  ...   
25%       5.000000  1954.000000   1967.000000     0.000000     0.000000  ...   
50%       5.000000  1973.000000   1994.000000     0.000000   383.500000  ...   
75%       6.000000  2000.000000   2004.000000   166.000000   712.250000  ...   
max       9.000000  2010.000000   2010.000000  1600.000000  5644.000000  ...   

        WoodDeckSF  OpenPorchSF  EnclosedPorch    3SsnPorch  ScreenPorch  \
count  1460.000000  1460.000000    1460.000000  1460.000000  1460.000000   
mean     94.244521    46.660274      21.954110     3.409589    15.060959   
std     125.338794    66.256028      61.119149    29.317331    55.757415   
min       0.000000     0.000000       0.000000     0.000000     0.000000   
25%       0.000000     0.000000       0.000000     0.000000     0.000000   
50%       0.000000    25.000000       0.000000     0.000000     0.000000   
75%     168.000000    68.000000       0.000000     0.000000     0.000000   
max     857.000000   547.000000     552.000000   508.000000   480.000000   

          PoolArea       MiscVal       MoSold       YrSold      SalePrice  
count  1460.000000   1460.000000  1460.000000  1460.000000    1460.000000  
mean      2.758904     43.489041     6.321918  2007.815753  180921.195890  
std      40.177307    496.123024     2.703626     1.328095   79442.502883  
min       0.000000      0.000000     1.000000  2006.000000   34900.000000  
25%       0.000000      0.000000     5.000000  2007.000000  129975.000000  
50%       0.000000      0.000000     6.000000  2008.000000  163000.000000  
75%       0.000000      0.000000     8.000000  2009.000000  214000.000000  
max     738.000000  15500.000000    12.000000  2010.000000  755000.000000  

[8 rows x 38 columns]

#### Responda las siguientes preguntas

In [ ]:
# ¿Cuál es el tamaño de lote promedio, redondeado al entero más cercano?
avg_lot_size = 10517

# Al día de hoy, que tan antigua es la casa más nueva (año actual - año en que fue construida la casa)
newest_home_age = 14

#### Realizando el primer modelo de aprendizaje automático

In [ ]:
melbourne_data.columns

Index(['Suburb', 'Address', 'Rooms', 'Type', 'Price', 'Method', 'SellerG',
       'Date', 'Distance', 'Postcode', 'Bedroom2', 'Bathroom', 'Car',
       'Landsize', 'BuildingArea', 'YearBuilt', 'CouncilArea', 'Lattitude',
       'Longtitude', 'Regionname', 'Propertycount'],
      dtype='object')

In [ ]:
melbourne_data.shape

(13580, 21)

In [ ]:
melbourne_data = melbourne_data.dropna(axis=0)

In [ ]:
melbourne_data.shape

(6196, 21)

#### Elegimos el valor a predecir

In [ ]:
y = melbourne_data.Price
print(y)

1        1035000.0
2        1465000.0
4        1600000.0
6        1876000.0
7        1636000.0
           ...    
12205     601000.0
12206    1050000.0
12207     385000.0
12209     560000.0
12212    2450000.0
Name: Price, Length: 6196, dtype: float64


#### Elegimos los atributos (features) que utilizaremos para intentar predecir el precio de las casas

In [ ]:
melbourne_features = ['Rooms', 'Bathroom', 'Landsize', 'Lattitude', 'Longtitude']

In [ ]:
X = melbourne_data[melbourne_features]

In [ ]:
X.head()

Rooms  Bathroom  Landsize  Lattitude  Longtitude
1      2       1.0     156.0   -37.8079    144.9934
2      3       2.0     134.0   -37.8093    144.9944
4      4       1.0     120.0   -37.8072    144.9941
6      3       2.0     245.0   -37.8024    144.9993
7      2       1.0     256.0   -37.8060    144.9954

#### Construimos el modelo

In [ ]:
from sklearn.tree import DecisionTreeRegressor

# Definimos el modelo. Especificamos un valor para random_state para asegurar los mismos resultados en cada ejecución
melbourne_model = DecisionTreeRegressor(random_state=1)

# Entrenamos el modelo con las variables independientes (X) y la variable dependiente o valor a predecir (y)
melbourne_model.fit(X, y)

DecisionTreeRegressor(random_state=1)

In [ ]:
print("Haciendo predicciones para las siguientes 5 casas:")
print(X.head())

Making predictions for the following 5 houses:
   Rooms  Bathroom  Landsize  Lattitude  Longtitude
1      2       1.0     156.0   -37.8079    144.9934
2      3       2.0     134.0   -37.8093    144.9944
4      4       1.0     120.0   -37.8072    144.9941
6      3       2.0     245.0   -37.8024    144.9993
7      2       1.0     256.0   -37.8060    144.9954


In [ ]:
y.head()

1    1035000.0
2    1465000.0
4    1600000.0
6    1876000.0
7    1636000.0
Name: Price, dtype: float64

In [ ]:
print("Las predicciones son:")
print(melbourne_model.predict(X.head()))

The predictions are
[1035000. 1465000. 1600000. 1876000. 1636000.]


#### Realizemos una predicción con datos completamente nuevos

In [ ]:
cuartos = int(input("Dame el número de cuartos: "))
banos = int(input("Dame el número de baños: "))
area = int(input("Dame el área: "))
latitud = float(input("Dame la latitud: "))
longitud = float(input("Dame la longitud: "))

Dame el número de cuartos: 5
Dame el número de baños: 3
Dame el área: 200
Dame la latitud: -37.8079
Dame la longitud: 144.9934


In [ ]:
dato = [cuartos, banos, area, latitud, longitud]
print(dato)

[5, 3, 200, -37.8079, 144.9934]


In [ ]:
valor = melbourne_model.predict([dato])

/usr/local/lib/python3.10/dist-packages/sklearn/base.py:439: UserWarning: X does not have valid feature names, but DecisionTreeRegressor was fitted with feature names
  warnings.warn(


In [ ]:
print(f"El valor de la casa es: {valor[0]}")

El valor de la casa es: 1212500.0


#### Creando un modelo con el dataset de `train_homedata`

In [ ]:
home_data.columns

Index(['Id', 'MSSubClass', 'MSZoning', 'LotFrontage', 'LotArea', 'Street',
       'Alley', 'LotShape', 'LandContour', 'Utilities', 'LotConfig',
       'LandSlope', 'Neighborhood', 'Condition1', 'Condition2', 'BldgType',
       'HouseStyle', 'OverallQual', 'OverallCond', 'YearBuilt', 'YearRemodAdd',
       'RoofStyle', 'RoofMatl', 'Exterior1st', 'Exterior2nd', 'MasVnrType',
       'MasVnrArea', 'ExterQual', 'ExterCond', 'Foundation', 'BsmtQual',
       'BsmtCond', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinSF1',
       'BsmtFinType2', 'BsmtFinSF2', 'BsmtUnfSF', 'TotalBsmtSF', 'Heating',
       'HeatingQC', 'CentralAir', 'Electrical', '1stFlrSF', '2ndFlrSF',
       'LowQualFinSF', 'GrLivArea', 'BsmtFullBath', 'BsmtHalfBath', 'FullBath',
       'HalfBath', 'BedroomAbvGr', 'KitchenAbvGr', 'KitchenQual',
       'TotRmsAbvGrd', 'Functional', 'Fireplaces', 'FireplaceQu', 'GarageType',
       'GarageYrBlt', 'GarageFinish', 'GarageCars', 'GarageArea', 'GarageQual',
       'GarageCond', 'PavedDrive

#### Seleccionamos la variable a predecir (y)

In [ ]:
#y = home_data.SalePrice
y = home_data["SalePrice"]

#### Seleccionamos las variables independientes (X)

In [ ]:
# Create the list of features below
feature_names = ["LotArea","YearBuilt","1stFlrSF","2ndFlrSF","FullBath","BedroomAbvGr","TotRmsAbvGrd"]

# Select data corresponding to features in feature_names
X = home_data[feature_names]

In [ ]:
X.describe()

LotArea    YearBuilt     1stFlrSF     2ndFlrSF     FullBath  \
count    1460.000000  1460.000000  1460.000000  1460.000000  1460.000000   
mean    10516.828082  1971.267808  1162.626712   346.992466     1.565068   
std      9981.264932    30.202904   386.587738   436.528436     0.550916   
min      1300.000000  1872.000000   334.000000     0.000000     0.000000   
25%      7553.500000  1954.000000   882.000000     0.000000     1.000000   
50%      9478.500000  1973.000000  1087.000000     0.000000     2.000000   
75%     11601.500000  2000.000000  1391.250000   728.000000     2.000000   
max    215245.000000  2010.000000  4692.000000  2065.000000     3.000000   

       BedroomAbvGr  TotRmsAbvGrd  
count   1460.000000   1460.000000  
mean       2.866438      6.517808  
std        0.815778      1.625393  
min        0.000000      2.000000  
25%        2.000000      5.000000  
50%        3.000000      6.000000  
75%        3.000000      7.000000  
max        8.000000     14.000000

In [ ]:
X.head()

LotArea  YearBuilt  1stFlrSF  2ndFlrSF  FullBath  BedroomAbvGr  \
0     8450       2003       856       854         2             3   
1     9600       1976      1262         0         2             3   
2    11250       2001       920       866         2             3   
3     9550       1915       961       756         1             3   
4    14260       2000      1145      1053         2             4   

   TotRmsAbvGrd  
0             8  
1             6  
2             6  
3             7  
4             9

#### Configuramos y entrenamos el modelo

In [ ]:
from sklearn.tree import DecisionTreeRegressor
#specify the model.
#For model reproducibility, set a numeric value for random_state when specifying the model
iowa_model = DecisionTreeRegressor(random_state=1)

# Fit the model
iowa_model.fit(X,y)

DecisionTreeRegressor(random_state=1)

#### Hacemos algunas predicciones

In [ ]:
#Usaremos solo los primeros 5 datos del dataframe de entrenamiento
predictions = iowa_model.predict(X.head())
print(predictions)

[208500. 181500. 223500. 140000. 250000.]


In [ ]:
y.head()

0    208500
1    181500
2    223500
3    140000
4    250000
Name: SalePrice, dtype: int64

### Validación del modelo

In [5]:
import pandas as pd

# Load data
melbourne_file_path = "https://raw.githubusercontent.com/vbatiz/intro-ML/main/notebooks/datasets/melb_data.csv"
melbourne_data = pd.read_csv(melbourne_file_path)
# Filter rows with missing price values
filtered_melbourne_data = melbourne_data.dropna(axis=0)
# Choose target and features

y = filtered_melbourne_data.Price

melbourne_features = ['Rooms', 'Bathroom', 'Landsize', 'BuildingArea',
                        'YearBuilt', 'Lattitude', 'Longtitude']

X = filtered_melbourne_data[melbourne_features]

from sklearn.tree import DecisionTreeRegressor
# Define model
melbourne_model = DecisionTreeRegressor()
# Fit model
melbourne_model.fit(X, y)

DecisionTreeRegressor()

In [8]:
filtered_melbourne_data.describe()

Rooms         Price     Distance     Postcode     Bedroom2  \
count  6196.000000  6.196000e+03  6196.000000  6196.000000  6196.000000   
mean      2.931407  1.068828e+06     9.751097  3101.947708     2.902034   
std       0.971079  6.751564e+05     5.612065    86.421604     0.970055   
min       1.000000  1.310000e+05     0.000000  3000.000000     0.000000   
25%       2.000000  6.200000e+05     5.900000  3044.000000     2.000000   
50%       3.000000  8.800000e+05     9.000000  3081.000000     3.000000   
75%       4.000000  1.325000e+06    12.400000  3147.000000     3.000000   
max       8.000000  9.000000e+06    47.400000  3977.000000     9.000000   

          Bathroom          Car      Landsize  BuildingArea    YearBuilt  \
count  6196.000000  6196.000000   6196.000000   6196.000000  6196.000000   
mean      1.576340     1.573596    471.006940    141.568645  1964.081988   
std       0.711362     0.929947    897.449881     90.834824    38.105673   
min       1.000000     0.000000      0.000000      0.000000  1196.000000   
25%       1.000000     1.000000    152.000000     91.000000  1940.000000   
50%       1.000000     1.000000    373.000000    124.000000  1970.000000   
75%       2.000000     2.000000    628.000000    170.000000  2000.000000   
max       8.000000    10.000000  37000.000000   3112.000000  2018.000000   

         Lattitude   Longtitude  Propertycount  
count  6196.000000  6196.000000    6196.000000  
mean    -37.807904   144.990201    7435.489509  
std       0.075850     0.099165    4337.698917  
min     -38.164920   144.542370     389.000000  
25%     -37.855438   144.926198    4383.750000  
50%     -37.802250   144.995800    6567.000000  
75%     -37.758200   145.052700   10175.000000  
max     -37.457090   145.526350   21650.000000

Calculamos el error absoluto medio (MAE, Mean Absolute Error)

In [6]:
from sklearn.metrics import mean_absolute_error

predicted_home_prices = melbourne_model.predict(X)
mean_absolute_error(y, predicted_home_prices)

434.71594577146544

#### Dividimos el conjunto de datos en entrenamiento y prueba para validar de mejor forma el modelo.

In [7]:
from sklearn.model_selection import train_test_split

# split data into training and validation data, for both features and target
# The split is based on a random number generator. Supplying a numeric value to
# the random_state argument guarantees we get the same split every time we
# run this script.
train_X, val_X, train_y, val_y = train_test_split(X, y, random_state = 0)
# Define model
melbourne_model = DecisionTreeRegressor()
# Fit model
melbourne_model.fit(train_X, train_y)

# get predicted prices on validation data
val_predictions = melbourne_model.predict(val_X)
print(mean_absolute_error(val_y, val_predictions))

259702.42156229826


### Validando un modelo con otro dataset

In [9]:
# Code you have previously used to load data
import pandas as pd
from sklearn.tree import DecisionTreeRegressor

# Path of the file to read
iowa_file_path = "https://raw.githubusercontent.com/vbatiz/intro-ML/main/notebooks/datasets/train_homedata.csv"
home_data = pd.read_csv(iowa_file_path)

y = home_data.SalePrice #Valor a predecir

feature_columns = ['LotArea', 'YearBuilt', '1stFlrSF', '2ndFlrSF', 'FullBath', 'BedroomAbvGr', 'TotRmsAbvGrd']
X = home_data[feature_columns]

# Specify Model
iowa_model = DecisionTreeRegressor()
# Fit Model
iowa_model.fit(X, y)

print("First in-sample predictions:", iowa_model.predict(X.head()))
print("Actual target values for those homes:", y.head().tolist())

First in-sample predictions: [208500. 181500. 223500. 140000. 250000.]
Actual target values for those homes: [208500, 181500, 223500, 140000, 250000]


1. Creamos los conjuntos de entrenamiento y prueba

In [14]:
from sklearn.model_selection import train_test_split

train_X, val_X, train_y, val_y = train_test_split(X, y, random_state = 1)

In [15]:
val_X.head()

LotArea  YearBuilt  1stFlrSF  2ndFlrSF  FullBath  BedroomAbvGr  \
258     12435       2001       963       829         2             3   
267      8400       1939      1052       720         2             4   
288      9819       1967       900         0         1             3   
649      1936       1970       630         0         1             1   
1233    12160       1959      1188         0         1             3   

      TotRmsAbvGrd  
258              7  
267              8  
288              5  
649              3  
1233             6

2. Creamos y entrenamos el modelo con los datos de entrenamiento

In [16]:
# You imported DecisionTreeRegressor in your last exercise
# and that code has been copied to the setup code above. So, no need to
# import it again

# Specify the model
iowa_model = DecisionTreeRegressor(random_state=1)

# Fit iowa_model with the training data.
iowa_model.fit(train_X,train_y)

DecisionTreeRegressor(random_state=1)

3. Realizamos predicciones con los datos de validación (estos datos no los conoce el modelo).

In [17]:
# Predict with all validation observations
val_predictions = iowa_model.predict(val_X)

In [18]:
# print the top few validation predictions
print(val_predictions[:5])
# print the top few actual prices from validation data
print(val_y.head())

[186500. 184000. 130000.  92000. 164500.]
258     231500
267     179500
288     122000
649      84500
1233    142000
Name: SalePrice, dtype: int64


4. Calculamos el Error Absoluto Medio (MAE) con los datos de validación

In [19]:
from sklearn.metrics import mean_absolute_error
val_mae = mean_absolute_error(val_y,val_predictions)

# uncomment following line to see the validation_mae
print(val_mae)

29652.931506849316


### Underfitting y Overfitting

Mejorando el modelo

In [22]:
# Data Loading Code Runs At This Point
import pandas as pd

# Load data
melbourne_file_path = "https://raw.githubusercontent.com/vbatiz/intro-ML/main/notebooks/datasets/melb_data.csv"
melbourne_data = pd.read_csv(melbourne_file_path)
# Filter rows with missing values
filtered_melbourne_data = melbourne_data.dropna(axis=0)
# Choose target and features
y = filtered_melbourne_data.Price
melbourne_features = ['Rooms', 'Bathroom', 'Landsize', 'BuildingArea',
                        'YearBuilt', 'Lattitude', 'Longtitude']
X = filtered_melbourne_data[melbourne_features]

from sklearn.model_selection import train_test_split

# split data into training and validation data, for both features and target
train_X, val_X, train_y, val_y = train_test_split(X, y,random_state = 0)

In [23]:
from sklearn.metrics import mean_absolute_error
from sklearn.tree import DecisionTreeRegressor

def get_mae(max_leaf_nodes, train_X, val_X, train_y, val_y):
    model = DecisionTreeRegressor(max_leaf_nodes=max_leaf_nodes, random_state=0)
    model.fit(train_X, train_y)
    preds_val = model.predict(val_X)
    mae = mean_absolute_error(val_y, preds_val)
    return(mae)

Probamos con diferentes valores para max_leaf_nodes

In [27]:
# compare MAE with differing values of max_leaf_nodes
#for max_leaf_nodes in [5, 50, 500, 5000]:
for max_leaf_nodes in range(380,501,10):
    my_mae = get_mae(max_leaf_nodes, train_X, val_X, train_y, val_y)
    print("Max leaf nodes: %d  \t\t Mean Absolute Error:  %d" %(max_leaf_nodes, my_mae))

Max leaf nodes: 380  		 Mean Absolute Error:  242472
Max leaf nodes: 390  		 Mean Absolute Error:  241811
Max leaf nodes: 400  		 Mean Absolute Error:  242906
Max leaf nodes: 410  		 Mean Absolute Error:  243616
Max leaf nodes: 420  		 Mean Absolute Error:  244037
Max leaf nodes: 430  		 Mean Absolute Error:  243300
Max leaf nodes: 440  		 Mean Absolute Error:  244168
Max leaf nodes: 450  		 Mean Absolute Error:  243036
Max leaf nodes: 460  		 Mean Absolute Error:  242777
Max leaf nodes: 470  		 Mean Absolute Error:  243543
Max leaf nodes: 480  		 Mean Absolute Error:  244300
Max leaf nodes: 490  		 Mean Absolute Error:  243530
Max leaf nodes: 500  		 Mean Absolute Error:  243495


El mejor resultado nos da con 390 nodos

#### Realizamos el ejercicio con el dataset de Iowa

In [28]:
# Code you have previously used to load data
import pandas as pd
from sklearn.metrics import mean_absolute_error
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeRegressor


# Path of the file to read
iowa_file_path = "https://raw.githubusercontent.com/vbatiz/intro-ML/main/notebooks/datasets/train_homedata.csv"

home_data = pd.read_csv(iowa_file_path)
# Create target object and call it y
y = home_data.SalePrice
# Create X
features = ['LotArea', 'YearBuilt', '1stFlrSF', '2ndFlrSF', 'FullBath', 'BedroomAbvGr', 'TotRmsAbvGrd']
X = home_data[features]

# Split into validation and training data
train_X, val_X, train_y, val_y = train_test_split(X, y, random_state=1)

# Specify Model
iowa_model = DecisionTreeRegressor(random_state=1)
# Fit Model
iowa_model.fit(train_X, train_y)

# Make validation predictions and calculate mean absolute error
val_predictions = iowa_model.predict(val_X)
val_mae = mean_absolute_error(val_predictions, val_y)
print("Validation MAE: {:,.0f}".format(val_mae))

Validation MAE: 29,653


In [32]:
candidate_max_leaf_nodes = [5, 25, 50, 95, 100, 101, 102, 103, 104,105, 250, 500]
# Write loop to find the ideal tree size from candidate_max_leaf_nodes
for nodos in candidate_max_leaf_nodes:
    print(f"Nodos: {nodos}: MAE: ",get_mae(nodos,train_X,val_X,train_y,val_y))

Nodos: 5: MAE:  35044.51299744237
Nodos: 25: MAE:  29016.41319191076
Nodos: 50: MAE:  27405.930473214907
Nodos: 95: MAE:  27268.728164951488
Nodos: 100: MAE:  27282.50803885739
Nodos: 101: MAE:  27233.421280866525
Nodos: 102: MAE:  27071.002102784332
Nodos: 103: MAE:  27062.12526574542
Nodos: 104: MAE:  27018.507087880665
Nodos: 105: MAE:  27171.94053388672
Nodos: 250: MAE:  27893.822225701646
Nodos: 500: MAE:  29454.18598068598


Observamos que el mejor valor nos da con 100 nodos

#### Entrenamos el modelo con los datos completos y el número de nodos seleccionado

In [33]:
# Fill in argument to make optimal size and uncomment
final_model = DecisionTreeRegressor(max_leaf_nodes=104)

# fit the final model and uncomment the next two lines
final_model.fit(X, y)

DecisionTreeRegressor(max_leaf_nodes=104)

In [34]:
predicciones = final_model.predict(val_X.head())
print("Predicciones:")
print(predicciones)
print("Valores reales:")
print(val_y.head())

Predicciones:
[182392.31944444 163142.5        125362.57333333  70167.58333333
 146415.0075188 ]
Valores reales:
258     231500
267     179500
288     122000
649      84500
1233    142000
Name: SalePrice, dtype: int64


### Random Forest

In [35]:
import pandas as pd

# Load data
melbourne_file_path = "https://raw.githubusercontent.com/vbatiz/intro-ML/main/notebooks/datasets/melb_data.csv"
melbourne_data = pd.read_csv(melbourne_file_path)
# Filter rows with missing values
melbourne_data = melbourne_data.dropna(axis=0)
# Choose target and features
y = melbourne_data.Price
melbourne_features = ['Rooms', 'Bathroom', 'Landsize', 'BuildingArea',
                        'YearBuilt', 'Lattitude', 'Longtitude']
X = melbourne_data[melbourne_features]

from sklearn.model_selection import train_test_split

# split data into training and validation data, for both features and target
# The split is based on a random number generator. Supplying a numeric value to
# the random_state argument guarantees we get the same split every time we
# run this script.
train_X, val_X, train_y, val_y = train_test_split(X, y,random_state = 0)

In [36]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_absolute_error

forest_model = RandomForestRegressor(random_state=1)
forest_model.fit(train_X, train_y)
melb_preds = forest_model.predict(val_X)
print(mean_absolute_error(val_y, melb_preds))

191669.7536453626


Revisamos con el dataset de Iowa

In [38]:
# Code you have previously used to load data
import pandas as pd
from sklearn.metrics import mean_absolute_error
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeRegressor


# Path of the file to read
iowa_file_path = "https://raw.githubusercontent.com/vbatiz/intro-ML/main/notebooks/datasets/train_homedata.csv"

home_data = pd.read_csv(iowa_file_path)
# Create target object and call it y
y = home_data.SalePrice
# Create X
features = ['LotArea', 'YearBuilt', '1stFlrSF', '2ndFlrSF', 'FullBath', 'BedroomAbvGr', 'TotRmsAbvGrd']
X = home_data[features]

# Split into validation and training data
train_X, val_X, train_y, val_y = train_test_split(X, y, random_state=1)

# Specify Model
iowa_model = DecisionTreeRegressor(random_state=1)
# Fit Model
iowa_model.fit(train_X, train_y)

# Make validation predictions and calculate mean absolute error
val_predictions = iowa_model.predict(val_X)
val_mae = mean_absolute_error(val_predictions, val_y)
print("Validation MAE when not specifying max_leaf_nodes: {:,.0f}".format(val_mae))

# Using best value for max_leaf_nodes
iowa_model = DecisionTreeRegressor(max_leaf_nodes=104, random_state=1)
iowa_model.fit(train_X, train_y)
val_predictions = iowa_model.predict(val_X)
val_mae = mean_absolute_error(val_predictions, val_y)
print("Validation MAE for best value of max_leaf_nodes: {:,.0f}".format(val_mae))


Validation MAE when not specifying max_leaf_nodes: 29,653
Validation MAE for best value of max_leaf_nodes: 27,019


Probamos con Random Forest

In [39]:
from sklearn.ensemble import RandomForestRegressor

# Define the model. Set random_state to 1
rf_model = RandomForestRegressor(random_state=1)

# fit your model
rf_model.fit(train_X,train_y)

# Realizar predicciones
val_predictions = rf_model.predict(val_X)

# Calculate the mean absolute error of your Random Forest model on the validation data
rf_val_mae = mean_absolute_error(val_predictions, val_y)

print("Validation MAE for Random Forest Model: {}".format(rf_val_mae))

Validation MAE for Random Forest Model: 21857.15912981083
